aiokafka/producer/message_accumulator.py

```diff
@@ -35,13 +36,27 @@ def append(self, *, timestamp, key, value):
         self._relative_offset += 1
         return actual_size
 
-    def _build(self):
-        assert not self._closed
+    def close(self):
+        if self._closed:
+            return
         self._closed = True
-        buffer = self._builder.build()
+        data = self._builder.build()
+        self._buffer = io.BytesIO(Int32.encode(len(data)) + data)
         del self._builder
 
-        return io.BytesIO(Int32.encode(len(buffer)) + buffer)
+    def _build(self):
+        if not self._closed:
+            self.close()
+        return self._buffer
+
+    def size(self):
+        if self._buffer:
+            return self._buffer.getbuffer().nbytes
```

Comment:
Why do you use such a strange way to get the size of the buffer? The creation of a memoryview is not so cheap. Why not call `self._buffer.tell()`? It should always be at the end anyway.

Reply:
No, it's not particularly cheap. That's why I originally had `BatchBuilder` tracking the size independently.

The buffer will only be at the end after the request is issued to the broker. During the time between `_build()` and the actual on-the-wire request creation in `producer._send_produce_req()` the buffer's position is 0. There are multiple yields to the event loop in that space--particularly if a batch is already in-flight.

The good news is that this method is user-facing and more likely to be called before submission of the batch than after, so most uses will use the cheaper `_builder.size()` call.

If you see this as an issue, we can save the size between calls after `_builder` is deleted so `getbuffer()` is only called once. Since this is a convenience method to the user whose use case is mostly pre-submission, I don't consider it a big deal.